### PR TITLE
Refactor and test `leaching_runoff_erosion.py` 3/3

### LINKED ISSUE #504: refactor and test leaching_runoff_erosion.py (soil/nitrogen_cycling) into SC_redesign directory

### PR DESCRIPTION
**Pull request by Isgotkowitz**
# Summary
Finishes the refactor of `leaching_runoff_erosion.py` and closes #504 

This PR is stacked on #487 

# Main Changes
- `leaching_runoff_erosion.py`
  - Now includes a top level method that calls the two helper methods that execute surface loss and leaching of nitrogen respectively.
  - Please pay special attention to `_leach_nitrogen()` and its test. This method is particularly complex because it requires doing calculations over the entire soil profile, saving the results, then applying those results over the entire soil profile.
  - Several `@staticmethod`s have been added as wrappers that calculate how much nitrogen is lost to a particular process, making it more convenient to execute the same process on different nitrogen pools.

### CODE REVIEW

**Review by morrowcj — APPROVED** [review]
Looks good to me. I really like the "Notes" sections that you've written. They help to clarify a lot.

File: SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/leaching_runoff_erosion.py [hunk 1]
```diff
@@ -20,26 +28,157 @@ def __init__(self, soil_data: Optional[SoilData], field_size: Optional[float] =
...
+            if layer.percolated_water == 0.0:
+                nitrogen_percolated_to_next_layer = {
+                    "nitrates": 0,
+                    "ammonium": 0,
+                    "active_organic": 0
+                }
+                percolated_nitrogen.append(nitrogen_percolated_to_next_layer)
+                continue
+
+            nitrate_extraction_coefficient = 1.0 if layer.top_depth == 0 else 2.5
```
Comment (matthew7838) [inline SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/leaching_runoff_erosion.py:133]: In the pseudocode, 2.5 is for layers 2 and 3 only, is the current implementation something that's discussed with the experts? I also don't see 2.5 being used in the original main routine.

File: SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/leaching_runoff_erosion.py [hunk 2]
```diff
@@ -263,3 +448,46 @@ def _determine_leached_nitrogen(nitrogen_concentration: float, percolation_amoun
...
+            The content of nitrogen in the given pool in the current layer of soil (kg / ha)
+        field_capacity_content : float
+            Amount of water in this soil layer when at field capacity (mm)
+        percolation_amount : float
+            Amount of water that percolated out of the current soil layer on this day (mm)
+        leaching_extraction_coefficient : float
+            Coefficient for adjusting the amount leached based on depth (unitless)
+        is_active_organic_nitrogen
+            Status indicating whether the pool being leached from is the active organic nitrogen pool or not
+            (True / False)
```
Comment (matthew7838) [inline SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/leaching_runoff_erosion.py]:
```suggestion
        is_active_organic_nitrogen  : bool
            Status indicating whether the pool being leached from is the active organic nitrogen pool or not
```

Comment (matthew7838) [inline SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/leaching_runoff_erosion.py]:
```suggestion
        is_active_organic_nitrogen  : bool
            Status indicating whether the pool being leached from is the active organic nitrogen pool or not
```

**Review by matthew7838 — APPROVED** [review]
Looks really good, this file's complexity way exceeded what I thought. I really liked how leached nitrogen is refactored, it's much more straightforward and the notes were helpful. Just a few suggestions and comments.

File: SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/leaching_runoff_erosion.py (hunk 1 above)
Comment (Isgotkowitz) [reply SC_redesign/Crop_and_Soil/soil/nitrogen_cycling/leaching_runoff_erosion.py:133]: For the SME take on what is going on here, I think we need to wait until post-MVP delivery, which we'll need to get anyway for #495. As long as were implementing what was in the old `NitrogenCycling` module we are good for now.

This part was implemented really poorly in the original code
```python
    # the coefficient of extraction for leaching is calibrated to 1.0
    # for layer 1 only
    layer.Cl = 1.0

    soil.soil_layers[0] = layer

    for layer in soil.soil_layers:

        # --- snip ---

        if percolation > 0:
            # --- snip ---

            # "pseudocode_soil" S.4.C.8
            NO3_percolation = NO3_percolation_conc * percolation / layer.Cl
            NH4_percolation = NH4_percolation_conc * percolation
            active_N_percolation = active_conc * percolation
```
`SoilLayer` objects had defaults of 2.5 for `CL`, so it does get used even though it is not clear.

**Pull request merged**